### PR TITLE
Restrict syntax of typed patterns

### LINKED ISSUE #10994: nested named irrefutable typetest pattern causes exhaustivity warning
## Minimized code

```Scala
def foo = true match
  case (b: Boolean): Boolean => ()
```

## Output

```scala
-- [E029] Pattern Match Exhaustivity Warning: sandbox/match/example.scala:1:10 -
1 |def foo = true match
  |          ^^^^
  |          match may not be exhaustive.
  |
  |          It would fail on pattern case: true, false

```

## Expectation

no warning.

the wildcard case produces no warning:
```scala
def foo = true match
  case (_: Boolean): Boolean => ()
```

### LINKED ISSUE #15893: Erratic compiler behavior for inlined function definitions
See [this](https://contributors.scala-lang.org/t/help-needed-teasing-apart-compiler-bugs/5880) discussion thread.

## Compiler version

3.1.3

## Minimized code

```Scala
sealed trait NatT
case class Zero() extends NatT
case class Succ[+N <: NatT](n: N) extends NatT

type Mod2[N <: NatT] <: NatT = N match
  case Zero => Zero
  case Succ[Zero] => Succ[Zero]
  case Succ[Succ[predPredN]] => Mod2[predPredN]

def mod2(n: NatT):  NatT = n match
  case Zero() => Zero()
  case Succ(Zero()) => Succ(Zero())
  case Succ(Succ(predPredN)) => mod2(predPredN)

inline def inlineMod2(inline n: NatT):  NatT = inline n match
  case Zero() => Zero()
  case Succ(Zero()) => Succ(Zero())
  case Succ(Succ(predPredN)) => inlineMod2(predPredN)

transparent inline def transparentInlineMod2(inline n: NatT):  NatT = inline n match
  case Zero() => Zero()
  case Succ(Zero()) => Succ(Zero())
  case Succ(Succ(predPredN)) => transparentInlineMod2(predPredN)

def dependentlyTypedMod2[N <: NatT](n: N): Mod2[N] = n match // exhaustivity warning; unexpected
  case Zero(): Zero => Zero()
  case Succ(Zero()): Succ[Zero] => Succ(Zero())
  case Succ(Succ(predPredN)): Succ[Succ[_]] => dependentlyTypedMod2(predPredN)

inline def inlineDependentlyTypedMod2[N <: NatT](inline n: N): Mod2[N] = inline n match
  case Zero(): Zero => Zero()
  case Succ(Zero()): Succ[Zero] => Succ(Zero())
  case Succ(Succ(predPredN)): Succ[Succ[_]] => inlineDependentlyTypedMod2(predPredN)

transparent inline def transparentInlineDependentlyTypedMod2[N <: NatT](inline n: N): Mod2[N] = inline n match
  case Zero(): Zero => Zero()
  case Succ(Zero()): Succ[Zero] => Succ(Zero())
  case Succ(Succ(predPredN)): Succ[Succ[_]] => transparentInlineDependentlyTypedMod2(predPredN)

def foo(n: NatT): NatT = mod2(n) match
  case Succ(Zero()) => Zero()
  case _ => n

inline def inlineFoo(inline n: NatT): NatT = inline inlineMod2(n) match
  case Succ(Zero()) => Zero()
  case _ => n

inline def transparentInlineFoo(inline n: NatT): NatT = inline transparentInlineMod2(n) match
  case Succ(Zero()) => Zero()
  case _ => n

@main def main(): Unit =
  println(mod2(Succ(Succ(Succ(Zero()))))) // prints Succ(Zero()), as expected
  println(foo(Succ(Succ(Succ(Zero()))))) // prints Zero(), as expected
  println(inlineMod2(Succ(Succ(Succ(Zero()))))) // prints Succ(Zero()), as expected
  println(inlineFoo(Succ(Succ(Succ(Zero()))))) // prints Succ(Succ(Succ(Zero()))); unexpected
  println(transparentInlineMod2(Succ(Succ(Succ(Zero()))))) // prints Succ(Zero()), as expected
  println(transparentInlineFoo(Succ(Succ(Succ(Zero()))))) // prints Zero(), as expected
  println(dependentlyTypedMod2(Succ(Succ(Succ(Zero()))))) // runtime error; unexpected
//  println(inlineDependentlyTypedMod2(Succ(Succ(Succ(Zero()))))) // doesn't compile; unexpected
//  println(transparentInlineDependentlyTypedMod2(Succ(Succ(Succ(Zero()))))) // doesn't compile; unexpected


```

## Output

```scala
Succ(Zero())
Zero()
Succ(Zero())
Succ(Succ(Succ(Zero())))
Succ(Zero())
Zero()
[error] java.lang.ClassCastException: Succ cannot be cast to Zero
[error]         at main$package$.dependentlyTypedMod2(main.scala:27)
[error]         at main$package$.main(main.scala:59)
[error]         at main.main(main.scala:52)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[error]         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[error]         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[error]         at java.lang.reflect.Method.invoke(Method.java:498) 
```

## Expectation

1. `inlineFoo(Succ(Succ(Succ(Zero()))))` should not return `Succ(Succ(Succ(Zero())))`; `inline` should not change function output.
2. Compilation should not be negatively affected by more refined types.
3. No runtime exception

### PR DESCRIPTION
**Pull request by dwijnand**
- Fix #10994: align typed pattern syntax to Scala 2
- Fixes #15893

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala [hunk 1]
```diff
@@ -2809,11 +2809,13 @@ object Parsers {
...
 
-    /**  Pattern1     ::= Pattern2 [Ascription]
+    /**  Pattern1     ::= PatVar Ascription
+     *                  | SimpleLiteral Ascription
+     *                  | Pattern2
      */
     def pattern1(location: Location = Location.InPattern): Tree =
       val p = pattern2()
-      if in.isColon then
+      if (isVarPattern(p) || p.isInstanceOf[Number]) && in.isColon then
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/parsing/Parsers.scala:2818]: The code does not match the syntax: simple literals can also be strings, characters, or booleans. Also there should be tests that all simple literals can be given a type ascription.

**Review by odersky — CHANGES_REQUESTED** [review]

File: compiler/src/dotty/tools/dotc/parsing/Parsers.scala (hunk 1 above)
Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/parsing/Parsers.scala:2819]: The numeric literal tests are in your genericNumLits, BigFloat, GenericNumLits tests.

**Review by bishabosha — APPROVED** [review]
approve for change to syntax documents

**Comment by dwijnand** [comment]
@odersky happy to go with this, allowing only numeric literals (in addition to pattern variables)?  Wouldn't want this to slip to 3.4.0.

**Review by odersky — COMMENTED** [review]
LGTM. But I am  not sure what happens with quotes and splices. Don't we need something like

  $x : T

? Is that still allowed? 

Reassigning to @nicolasstucki to get his input.

**Comment by dwijnand** [comment]
It is, like

```scala
x match
  case '{ $x: T } => // match using outer `T`
```

in tests/pos-macros/i10050.scala.

**Comment by odersky** [comment]
OK, then it's good to go.

**Comment by smarter** [comment]
This broke the CI, but @dwijnand is currently looking into it: https://github.com/lampepfl/dotty/pull/16361

**Comment by WojciechMazur** [comment]
The following snippet fails in nightly. Bisect points to this PR. Was that behaviour expected?:
```scala
@main def Test = 
    val x: Any = 1
    x match
      case x @ (works: Long) => x
      case x @ fails: Int => x 
```
errror
```scala
[error] bisect/test.scala:5:21: '=>' expected, but ':' found
[error]       case x @ fails: Int => x 
[error]         
```
The snippet is based on a code found in tasty-query by Open CB

**Comment by dwijnand** [comment]
Yes, that's expected.  And it's just a test case (https://github.com/scalacenter/tasty-query/blob/main/test-sources/src/main/scala/simple_trees/Bind.scala)

**Pull request merged**